### PR TITLE
Tighten the public API surface: root exports, listen types, dead-code removal

### PR DESCRIPTION
**Pull request by maxisbey**
Cleanup pass over the v2 public surface: a few missing exports, a few dead ones, and some no-op aliases.

## Motivation and Context

- **Root `mcp`**: exports `MCPServer` next to the existing `Client`, so `from mcp import MCPServer, Client` works (the `first-steps.md` tip claiming otherwise is removed). Also exports `NoBackChannelError`, which every server-initiated request path can raise on a 2026-07-28 connection. Drops the v1-carryover `SamplingRole` alias (use `mcp.types.Role`).
- **`mcp.client`**: re-exports the types `Client.listen()` yields and raises — `Subscription`, `SubscriptionLost`, `ListenNotSupportedError` — which previously required reaching into `mcp.client.subscriptions`.
- **Dead surface removed**: `StreamableHTTPError` / `ResumptionError` (never raised from a code path a caller could catch) and the unused `StreamableHTTPTransport.get_session_id()`; the no-op `RegistrationRequest` / `TokenSuccessResponse` auth-handler aliases; the `validate_extension_identifier` re-export alias in `mcp.server.extension`; the unreachable `__main__` guard in `mcp/cli/__init__.py`.
- The lowlevel server `Context` is not constructed by `ServerRunner` yet, so it moves to the private `mcp.server._context` module until it's wired back in.
- `_handle_resumption_request` now takes the resumption token as a required argument instead of re-deriving it, which is what let the unreachable no-token branch (and the two exception classes) go away.
- Enables ruff `RUF022` and sorts every `__all__`; `mcp_types/__init__.py` is exempted since its `__all__` is grouped by section comments on purpose.

## How Has This Been Tested?

Full `./scripts/test` (100% coverage held, `strict-no-cover` clean), plus a smoke test of the public boundary in a fresh interpreter: root/`mcp.client` imports resolve, removed names raise `ImportError`, and `from mcp import MCPServer, Client` round-trips a tool call both in-memory and over Streamable HTTP.

## Breaking Changes

Yes, all documented in `docs/migration.md`: `SamplingRole` is gone from the root package, `StreamableHTTPError`/`ResumptionError`/`get_session_id()` are gone from `mcp.client.streamable_http`, and `mcp.server.context.Context` is now private (`ServerRequestContext` and the middleware types are unchanged).

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

## Checklist
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context

<sub>[AI Disclaimer](https://gist.github.com/maxisbey/6123d132484e4c533eab519a2800693d)</sub>

### CODE REVIEW

File: docs/migration.md [hunk 1]
```diff
@@ -2771,7 +2780,7 @@ The user-facing methods for these features now carry `typing_extensions.deprecat
 
 - Sampling: `ServerSession.create_message()`, `ClientPeer.sample()`
 - Roots: `ServerSession.list_roots()`, `ClientPeer.list_roots()`, `ClientSession.send_roots_list_changed()`, `Client.send_roots_list_changed()`
-- Logging: `ServerSession.send_log_message()`, `Connection.log()`, `ClientSession.set_logging_level()`, `Client.set_logging_level()`, `mcp.server.context.Context.log()` (the lowlevel `Context`), and the `MCPServer` `Context` helpers `log()`, `debug()`, `info()`, `warning()`, `error()`
+- Logging: `ServerSession.send_log_message()`, `Connection.log()`, `ClientSession.set_logging_level()`, `Client.set_logging_level()`, and the `MCPServer` `Context` helpers `log()`, `debug()`, `info()`, `warning()`, `error()`
```
[the remark this answers is not in the document]
Comment (maxisbey) [reply docs/migration.md:2783]: migration.md is for breaking changes from v1 to v2. this class wasn't in v1 so there's no reason for it to be in migration.md
